Add jsonmod module

djmattyg007 opened the pull request:
This module lets you modify part of a JSON file by specifying the path
to the node you want to update, without caring about how the JSON file
is formatted. This is a more robust solution for updating parts of a
JSON document than using the lineinfile module, for example.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
jsonmod

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
As an example, assume you have a file containing this JSON document:

```json
    {
        "droids": [
            {
                "name": "C-4PO",
                "class": "protocol"
            },
            {
                "name": "R2-D2"
            }
        ]
    }
```

And I then ran a task that looked like so over this JSON file:

```yaml
    - name: Fix issues with droid information
      jsonmod:
        data_str: "{{ item.data }}"
        json_path: "{{ item.path }}"
        path: "/path/to/droids.json"
      with_items:
        - { data: "C-3PO", path: "droids[0]name" }
        - { data: "astromech", path: "droids[1]class" }
```

That document would now look like this:

```json
    {
        "droids": [
            {
                "name": "C-3PO",
                "class": "protocol"
            },
            {
                "name": "R2-D2",
                "class": "astromech"
            }
        ]
    }
```

The `jsonmod` module can also add actual JSON:

```yaml
    - name: Add info about BB-8
      jsonmod:
        data_json: {"name": "BB-8", "class": "astromech"}
        json_path: "droids[2]"
        path: "/path/to/droids.json"
```

The result would be:

```json
    {
        "droids": [
            {
                "name": "C-3PO",
                "class": "protocol"
            },
            {
                "name": "R2-D2",
                "class": "astromech"
            },
            {
                "name": "BB-8",
                "class": "astromech"
            }
        ]
    }
```

Finally, it can also remove nodes, like so:

```yaml
    - name: We don't need class data now
      jsonmod:
        json_path: "droids[{{ item }}]class"
        path: "/path/to/droids.json"
        remove: yes
      with_sequence: start=0 end=2
```

The result would be:

```json
    {
        "droids": [
            { "name": "C-3PO" },
            { "name": "R2-D2" },
            { "name": "BB-8" }
        ]
    }
```

My rationale for this is quite similar to what the `lineinfile` module is commonly used for. The difference is that JSON is a serialized format and therefore you can't make assumptions about what the file looks like.

djmattyg007 committed:
Add jsonmod module

This module lets you modify part of a JSON file by specifying the path
to the node you want to update, without caring about how the JSON file
is formatted. This is a more robust solution for updating parts of a
JSON document than using the lineinfile module, for example.

bcoca commented:
i would just read the file and use `|from_json` to make it into a data structure and `to_json` in a template to write it back out

djmattyg007 commented:
Sorry, I'm not sure how that matches what this module does? How can you programatically update a specific node within a (potentially huge) JSON document like that?

georgepsarakis commented:
I believe this functionality can be also achieved with a simple custom [Jinja filter](http://docs.ansible.com/ansible/playbooks_filters.html).

djmattyg007 commented:
Would you be able to provide an example?

georgepsarakis commented:
Well for dictionaries the [combine](http://docs.ansible.com/ansible/playbooks_filters.html#combining-hashes-dictionaries) filter already exists:

```
- name: Modify JSON
  copy:
    content: "{{ json_string | from_json | combine({"droids": {"name": "A", "class": "B"}}, recursive=True) | to_json }}"
```

If you want to modify lists, I think you will need to create something similar that needs to differentiate between lists and dictionaries in a nested structure.

djmattyg007 commented:
That looks like it would get very messy for nested structures. Most of the time it would require trial and error, and would not be easy to debug or modify.

djmattyg007 commented:
Also where do you get the JSON string from? How does this work with files on the remote host?

djmattyg007 commented:
Also what if you wanted to remove an element rather than add or replace elements?

georgepsarakis commented:
I understand your concerns, but can you mention some use cases for this? Because I don't think it is very frequent that someone modifies JSON files remotely as you describe it. Either, you have a template or somehow it is dynamically generated.

The `lineinfile` module has a lot of use cases because it is a very frequent need to interact with various system configuration files, which sometimes have defined their own format and perhaps it would be wrong to replace using your own template. For example `/etc/hosts`, `/etc/sudoers` have their own formatting rules, so you need something generic in order to edit the file in-place.

djmattyg007 commented:
You should think of this module as the JSON equivalent of.the `ini_file` module. The use case is simple: I want to be able to deterministically and idempotently update configuration files that take the form of JSON documents.

mattclay commented:
CI failure due to python 2.4 syntax error:

```
2017-02-01 18:23:41 Compiling ./lib/ansible/modules/files/jsonmod.py ...
2017-02-01 18:23:41   File "./lib/ansible/modules/files/jsonmod.py", line 304
2017-02-01 18:23:41     with open(filename, "r") as json_file:
2017-02-01 18:23:41             ^
2017-02-01 18:23:41 SyntaxError: invalid syntax
```

djmattyg007 commented:
I have no means of testing any code on python2.4 and have no idea what syntax limitations are imposed by this. Are you able to provide more details?

mattclay commented:
Python 2.4 doesn't support context managers, so `with` isn't available. Instead, you can use `try` .. `finally` to open and then close the file:

```python
try:
    # open file here
finally:
    # close file here
```

mattclay commented on lib/ansible/modules/files/jsonmod.py:
```diff
@@ -0,0 +1,370 @@
+
+from collections import OrderedDict
+import copy
+import json
```
This will also need to be updated for python 2.4 (thanks @abadger for pointing this out):

```python
try:
    import json
except ImportError:
    import simplejson as json
```

djmattyg007 committed:
Resolve various issues from review of jsonmod module

- Modify code to support python2.4
- Fix issues with actually writing JSON file to disk
- Update exception handling to support python3 and python2.4
- Remove trailing whitespace on one line

djmattyg007 commented:
I've pushed a new update that makes the requested changes, and rebased it against the current devel branch.

djmattyg007 commented:
It looks like the only remaining issue is the exception handling on python3.5. Is this something I can handle?

mattclay commented:
@djmattyg007 Here's an example of how to handle exceptions:

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/files/file.py#L286

mattclay commented:
CI failure due to PEP 8 issue:

```
2017-02-19 01:29:02 ERROR: PEP 8: lib/ansible/modules/files/jsonmod.py:250:1: W293 blank line contains whitespace (current)
```

You can run the PEP 8 tests locally with `make pep8`.

EvanK commented:
This is an interesting (albeit narrow use) module. I vote to shipit

dagwieers commented:
Sorry to come in quite late in the game. This module comes very close to another project named [ansible-xml](https://github.com/cmprescott/ansible-xml). This project has the intention to get merged upstream.

It would be nice if both the xml module and this json module have similar capabilities and parameter names (where they make sense). For XML you have XPath to do very complex manipulations, JSON has JSONPath. It would be nice if this module can work similarly in nature.

djmattyg007 commented:
While XPath may be a good fit for adding/updating/removing portions of an XML document, I don't believe JSONPath is a good fit for adding/updating/removing portions of a JSON document. Both XPath and JSONPath are tools for extracting data from XML and JSON documents respectively, not for modifying them.

I'm not sure how would you go about creating a system that automatically creates intermediary nodes (and hiding all of the complexity around that) without creating your own JSONPath-like parser that can deal with missing intermediary nodes, and working out just what intermediary nodes need to be created.

Additionally, what would you do in the case of an expression that contains wildcards or something similiar (that is designed to return multiple nodes when used as a query) when you also need to create intemediary nodes? I'm not sure of a way to achieve this without also creating my own JSONPath parser, which is not something I have the time to do.

djmattyg007 commented:
The only benefit I can see to introducing a dependency on the `jsonpath-rw` module is that it would mean this module wouldn't have to support python2.4, due to `jsonpath-rw` only supporting python2.6.

djmattyg007 commented:
Is there anything blocking this from inclusion in the next release of Ansible?

djmattyg007 commented:
ready_for_review

maxamillion commented on lib/ansible/modules/files/jsonmod.py:
```diff
@@ -0,0 +1,382 @@
+#!/usr/bin/python -tt
```
Shebang line should be `#!/usr/bin/python` as per [Contributing Modules Checklist](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_checklist.html#contributing-modules-checklist)

maxamillion commented on lib/ansible/modules/files/jsonmod.py:
```diff
@@ -0,0 +1,382 @@
+    - Used to update part of a JSON file by specifying a path to the node
+      you want to update and what you want to add to or replace it with.
+      It works regardless of how the JSON file is formatted.
+version_added: "1.0"
```
This needs to be `2.5` or `2.6` depending on when the changes come in for the merge window. If you can get the changes in today, we should be good for the `2.5` merge window (apologies for the late review).

djmattyg007 replied:
I can get it done in about 12 hours time, is that okay?

maxamillion commented:
needs_revision

bendoh commented:
I think this is a good idea, albeit a very narrow use case. I can think of good uses of JSON manipulation in playbooks and this could make it really easy and obvious.

I'm slightly concerned about the introduction of a new syntax for referencing and manipulating JSON structures that is not compatible with existing JS syntax. Since JSON is JavaScript Object Notation, this referencing syntax should also be somewhat compatible with it.

Where this really breaks for me is in the array-index-property syntax, `foo[1]bar`, which if it were `foo[1].bar`, would actually be interpretable by JS as well. Additionally, there doesn't appear to be support for arbitrary strings for object keys, which is also supported by JSON.

```
this.array["this is a string key with spaces and escaped dquotes  \"  "]
``` 

Additionally, I think a pretty darn common use case would to _append to_ or _pop from_ an array within the JSON structure, i.e., add to a list of entries. There doesn't appear to be a way to do this with the current proposal.

What if you introduced some syntax such as,

`the.array[]` to push to the end of the array (accept the `[]` syntax _only_ at the end of the path)

`this.array[]` with `remove: yes` from the end of the array

This isn't strictly compatible with PHP but it gives a way to reference the "end of the array" for adding or removing data. 

Similarly, I think there needs to be a way to _insert_ data into an array. This could be accomplished with an `insert:` option,

```
jsonmod:
  data_json: "fudge"
  json_path: "param1[0]"
  path: "/path/to/my.json"
  insert: yes
```

Before:
`{ "param1": [ "cookies", "cake" ] }`

After:
`{ "param1": [ "fudge", "cookies", "cake" ] }`

```
jsonmod:
  data_json: "caramel"
  json_path: "param1[2]"
  path: "/path/to/my.json"
  insert: yes
```

Before:
`{ "param1": [ "fudge", "cookies", "cake" ] }`

After:
`{ "param1": [ "fudge", "cookies", "caramel", "cake" ] }`

This will allow for the basic array operations: push, pop, shift, unshift, and then insert

maxamillion commented:
@bendoh you bring up a really good point about the introduction of a new notation for JSON manipulation, maybe this needs to be reworked a little.

bendoh commented:
LOL i just realized this PR is over a year old. Maybe it will heat back up!

maxamillion commented:
@bendoh yeah, I'm trying to work through the old backlog ... it's pretty long in the tooth.

ssbarnea commented:
This is clearly an important missing link in ansible and I hope this will go in sooner. The need to alter json files is real, tons of tools to keep their configuration in json format.

djmattyg007 commented:
So I've gotta be honest, I don't care about getting this merged now. The way this PR was treated was absolutely horrendous, and I feel no real desire to contribute anything further to ansible as a result. If the ansible maintainers actually want this merged, they'll have to bring it up to modern standards (documentation, etc) themselves.

All of the ludicrous off-topic discussions about this module from the very beginning have been incredibly frustrating to read and respond to. Crap like "just use JSON path" or "just use a jinja filter" completely miss the whole point. "Change the syntax so paths can also be valid javascript" is a massive WTF - why is this even desirable?? We're writing python here, not javascript. "Can you make it do all of these other things before it gets merged?" is an insult to my time, both past and future.

In the time since I last posted here, it seems another JSON-patching module was submitted for inclusion and given similar treatment:

https://github.com/ansible/ansible/pull/52931

It is extremely disappointing to discover that the governance of the Ansible project has not changed in the last three years.

For anyone hoping for a resolution to this, for future reference it appears that the author of #52931 has posted the json_patch module in a standalone repository:

https://github.com/particledecay/ansible-jsonpatch

Good luck to anyone who does attempt to contribute to the ansible project in future.

sivel commented:
Hi!

Thank you very much for your submission to Ansible. It sincerely means a lot to us that you've taken the time to contribute.

Per recent announcements at AnsibleFest Atlanta 2019, going forward we are no longer accepting new one off modules/plugins, or modules/plugins that create a new category.

However, we recommend looking into providing this functionality through Ansible Galaxy via Ansible Collections. You can find more information about collections at https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html.

Thank you once again for this and your interest in Ansible!

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

djmattyg007 commented:
Amazing. Despite me writing a clear complaint referencing another PR that was shutdown with a boilerplate response, you provide the same boilerplate response to a PR that was first opened long before that decision was made. Nice work.
